libcudacxx/include/cuda/std/__utility/unreachable.h

```diff
@@ -22,20 +22,34 @@
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-_CCCL_NORETURN _LIBCUDACXX_HIDE_FROM_ABI void __libcpp_unreachable()
+_CCCL_NORETURN _LIBCUDACXX_HIDE_FROM_ABI void unreachable()
 {
-  _CCCL_UNREACHABLE();
+#if defined(_CCCL_COMPILER_MSVC)
+#  define _CCCL_HOST_UNREACHABLE_IMPL() __assume(0)
+#else // ^^^ _CCCL_COMPILER_MSVC ^^^ / vvv !_CCCL_COMPILER_MSVC vvv
+#  define _CCCL_HOST_UNREACHABLE_IMPL() __builtin_unreachable()
+#endif // !_CCCL_COMPILER_MSVC
+
+#if defined(_CCCL_CUDA_COMPILER_CLANG)
+#  define _CCCL_DEVICE_UNREACHABLE_IMPL() __builtin_unreachable()
+#elif defined(__CUDA_ARCH__)
+#  if _CCCL_CUDACC_BELOW(11, 2)
+#    define _CCCL_DEVICE_UNREACHABLE_IMPL() __trap()
+#  elif _CCCL_CUDACC_BELOW(11, 3)
+#    define _CCCL_DEVICE_UNREACHABLE_IMPL() __builtin_assume(false)
+#  else
+#    define _CCCL_DEVICE_UNREACHABLE_IMPL() __builtin_unreachable()
+#  endif // CUDACC above 11.4
+#else // ^^^ __CUDA_ARCH__ ^^^ / vvv !__CUDA_ARCH__ vvv
+#  define _CCCL_DEVICE_UNREACHABLE_IMPL() __builtin_unreachable()
+#endif // !__CUDA_ARCH__
+
+  NV_IF_ELSE_TARGET(NV_IS_HOST, _CCCL_HOST_UNREACHABLE_IMPL();, _CCCL_DEVICE_UNREACHABLE_IMPL();)
+
+#undef _CCCL_HOST_UNREACHABLE_IMPL
+#undef _CCCL_DEVICE_UNREACHABLE_IMPL
```

Comment:
Th underlying issue is that one cannot use conditional compilation inside a `NV_IF_TARGET` because that itself is a macro

Reply:
Sure, this should be fixed now